metadetect/metadetect.py

```diff
@@ -115,57 +201,261 @@ def go(self):
+        ]
+        final_res = np.zeros(newres.shape[0], dtype=new_dt)
+        for c in final_res.dtype.names:
+            final_res[c] = newres[c]
+
+        return final_res
+
+    def _run_fitter_mbobs_sep(self, mbobs_list, nonshear_mbobs_list=None):
+        # run the fits
+        band_res = []
+        all_is_shear_band = []
+
+        for band in range(self.nband):
+            all_is_shear_band.append(True)
+            band_mbobs_list = []
+            for mbobs in mbobs_list:
+                _mbobs = ngmix.MultiBandObsList()
+                _mbobs.meta = copy.deepcopy(mbobs.meta)
+                _mbobs.append(mbobs[band])
+                band_mbobs_list.append(_mbobs)
+            band_res.append(self._fitter.go(band_mbobs_list))
+
+        if nonshear_mbobs_list is not None:
+            for band in range(self.nonshear_nband):
+                all_is_shear_band.append(False)
+                band_mbobs_list = []
+                for mbobs in nonshear_mbobs_list:
+                    _mbobs = ngmix.MultiBandObsList()
+                    _mbobs.meta = copy.deepcopy(mbobs.meta)
+                    _mbobs.append(mbobs[band])
+                    band_mbobs_list.append(_mbobs)
+                band_res.append(self._fitter.go(band_mbobs_list))
+
+        # combine the data by inverse variance weighted averages
+        tot_nband = self.nband + self.nonshear_nband
+        n = Namer(front=self['model'])
+        dt = [
+            ("flags", 'i4'),
+            (n("flags"), 'i4'),
+            (n("s2n"), "f8"),
+            (n("T"), "f8"),
+            (n("T_err"), "f8"),
+            (n("g"), "f8", 2),
+            (n("g_cov"), "f8", (2, 2)),
+            ('psf_g', 'f8', 2),
+            ('psf_T', 'f8'),
+            (n("T_ratio"), "f8"),
+        ]
+        if tot_nband > 1:
+            dt += [
+                (n("band_flux"), "f8", tot_nband),
+                (n("band_flux_err"), "f8", tot_nband),
+            ]
+        else:
+            dt += [
+                (n("band_flux"), "f8"),
+                (n("band_flux_err"), "f8"),
+            ]
+
+        res = np.zeros(len(mbobs_list), dtype=dt)
+        for ind in range(len(mbobs_list)):
+            # extract the wgts and band results
+            wgts = []
+            all_bres = []
+            for i, obslist in enumerate(mbobs_list[ind]):
+                msk = obslist[0].weight > 0
+                if not np.any(msk):
+                    raise RuntimeError("No non-zero weights for a band in metadetect!")
+                wgts.append(np.median(obslist[0].weight[msk]))
+                all_bres.append(band_res[i][ind:ind+1])
+
+            if nonshear_mbobs_list is not None:
+                for i in range(self.nband, tot_nband):
+                    all_bres.append(band_res[i][ind:ind+1])
+                    wgts.append(1)
+
+            wgts = np.array(wgts)
+            wgts[0:self.nband] = wgts[0:self.nband] / np.sum(wgts[0:self.nband])
```

Comment:
this looks confusing, as if there are extra data beyond the end which is not being rescaled, but I think that's not the case.

Reply:
There is extra data off the end if we have bands for fluxes. I have added a comment on this.